Integrated MeteorDemod into rtl_fm part of receive_meteor.sh

MihajloPi opened the pull request:
I've integrated MeteorDemod into an rtl_fm section of `receive_meteor.sh`. PLEASE TEST THOROUGLY!

I've never worked with Ansible before but I've edited its code since I THINK this would work. I've done similar small code edits of Ansible before myself and they worked, but I cannot stress enough of the importance to check my Ansible edits THOROUGLY! Ansible only needs to copy `settings.ini` file to `~/.config/meteordemod` and change latitude and longtitude coordinates in it to match receiver's location.

Those changes worked for me on buster and bullseye versions of the image. I only haven't added the Ansible part to my bullseye version on the Pi so, again, please test it. This is a work in progress and just my small contribution to this community.

Commit:
Create meteor_demodulate_qpsk.sh

The program used alias for a program which is not installed on Raspberry. Probably used to be on Raspberry as part of the script but got removed as newer programs were integrated.

Commit:
Added alias for MeteorDemod

I've added a new alias for MeteorDemod program created by Digitelektro: https://github.com/Digitelektro/MeteorDemod
I've also removed alias for meteor_demod since that program is no longer used.

Commit:
Added MeteorDemod to RTL_FM section

I've added MeteorDemod to RTL_FM section. TO BE TESTED THOROUGLY!

Commit:
Changed from GNU Radio to rtl_fm for Meteor M2

I've changed Meteor M2 reception method to be rtl_fm. That's until I integrate MeteorDemod to GNU Radio. I've only had it succesfully integrated it to rtl_fm section.

Commit:
Changed image names for Meteor

I've changed image names for Meteor. As a result, I've had to change which image names PHP has to look for in a database to display.

Commit:
Update Pass.php

Commit:
Changed image names for Meteor and NOAA

I've changed image names for Meteor and NOAA so that if it's daylight pass, the page displays image in visible light. At night, it displays IR image.

Commit:
Changed image names for Meteor and NOAA

I've changed image names for Meteor and NOAA so that if it's daylight pass, the page displays image in visible light. At night, it displays IR image.

Commit:
Added installation for MeteorDemod

I've added a way to install MeteorDemod. TO BE TESTED THOROUGLY!

Commit:
Added MeteorDemod .deb binary package

I've added a .deb binary for Meteor Demod provided by DigitElektro at: https://github.com/Digitelektro/MeteorDemod/releases/tag/2.0.0
I've also added a custom setting.ini file for MeteorDemod which works well.

Commit:
Change latitude and longtitude for MeteorDemod settings.ini file

I've added a latitude and longtitude variable for ansible to change in settings.ini file. I'm not familiar with Ansible so someone please correct my change. This file needs to be copied to `~/.config/meteordemod` so that folder has to have a file called settings.ini
I've added simple move command in `install_and_upgrade.sh` script to move that file to `~/.config/meteordemod`, but someone would have to manually input the correct coordinates every time someone runs `install_and_upgrade.sh` script since that would overwrite the file each time.

Commit:
Rename settings.ini to settings.ini.j2

Commit:
Copy MeteorDemod settings.ini file

I've added what I think WOULD work using Ansible to copy MeteorDemod settings.ini file into `~/.config/meteordemod` folder.

Again I'm not familiar with Ansible so please review this change and test it THOROUGLY!

Commit:
Update install_and_upgrade.sh

Commit:
Delete settings.ini

Cadair commented on ansible/roles/common/templates/settings.ini.j2:
```diff
@@ -0,0 +1,64 @@
+[Program]
```
Can we name this file something like `meteor_demod_settings.ini.j2" to make it more obvious what it is?

MihajloPi replied:
We can, it just needs to be called settings.ini when it is moved into `~/.config/meteordemod/`.

Cadair replied:
That is controlled by the `dest:` block in the ansible template directive, so that will happen :+1:

Cadair commented on config/settings.yml.sample:
```diff
@@ -25,12 +25,12 @@ test_enable_bias_tee: true
 #     **WARNING**: 'gnuradio' does not work with certain SDR devices (e.g. it will not currently
 #                  work with a RTL-SDR v3 dongle, as no image will be decoded from the bitstream)
-meteor_receiver: 'gnuradio'
 noaa_receiver: 'gnuradio'
+meteor_receiver: 'rtl_fm'
```
We don't want to merge this with this default set. We made the choice to move to gnuradio by default, and I don't think we want to change that now.

MihajloPi replied:
I did that since I've only so far integrated MeteorDemod into rtl_fm section since it seemed easier to me, I'll also integrate it into GNU Radio section. This is only for now.

Cadair commented on scripts/receive_meteor.sh:
```diff
@@ -176,59 +176,50 @@ if [ "$METEOR_RECEIVER" == "rtl_fm" ]; then
+  counter=1
+  for i in *.jpg
+    do
+      ${IMAGE_PROC_DIR}/meteor_normalize_annotate.sh "$i" "$i" 100 >> $NOAA_LOG 2>&1
```
Is the normalisation step this script performs still required when using meteordemod?

MihajloPi replied:
Well, it annotates the received Meteor image.

Cadair replied:
Oh I thought it also did another thing, my bad.

Cadair commented on scripts/receive_meteor.sh:
```diff
@@ -176,59 +176,50 @@ if [ "$METEOR_RECEIVER" == "rtl_fm" ]; then
-    rm "${AUDIO_FILE_BASE}.bmp"
-    rm "${AUDIO_FILE_BASE}.dec"
+      push_file_list="$push_file_list ${IMAGE_FILE_BASE}-${counter}-122-rectified.jpg "
+      ((counter++))
```
Why use counter and not `$i`?

MihajloPi replied:
Since MeteorDemod creates anywhere from 4 to 6 images and counter just creates new names that have the same base, but only differ in that counter number.

Cadair replied:
but doesn't counter always have the same value as `$i` (maybe `$i + 1`)? you are looping with a variable `$i` and then incrementing the counter at the end of each loop.

MihajloPi replied:
You didn't understand my way of thinking. $i holds image name, $counter is just an ordinal number which is appended to image filenames. That way, I can differentiate images from the same pass. It's the same with NOAA. NOAA appends enchamcements like -MSA, -MCIR etc. while Meteor appends -1, -2 etc.

Cadair replied:
oh I see, then may I suggest you change the variable name from `$i` to something more descriptive like `$imagename` my brain saw `i` and thought integer.

Cadair commented on scripts/receive_meteor.sh:
```diff
@@ -176,59 +176,50 @@ if [ "$METEOR_RECEIVER" == "rtl_fm" ]; then
       if [ "$ENABLE_EMAIL_PUSH" == "true" ]; then
         log "Emailing image" "INFO"
-        ${PUSH_PROC_DIR}/push_email.sh "${EMAIL_PUSH_ADDRESS}" "${IMAGE_FILE_BASE}-122-rectified.jpg" "${push_annotation}" >> $NOAA_LOG 2>&1
+	${PUSH_PROC_DIR}/push_email.sh "${EMAIL_PUSH_ADDRESS}" "${IMAGE_FILE_BASE}-2-122-rectified.jpg" "${push_annotation}" >> $NOAA_LOG 2>&1
```
This indentation change is incorrect?

MihajloPi replied:
What do you mean? I changed the filename.

Cadair replied:
According to these diffs the indentation at the start of the line has also smaller by 4 spaces.

Cadair commented on scripts/receive_meteor.sh:
```diff
@@ -176,59 +176,50 @@ if [ "$METEOR_RECEIVER" == "rtl_fm" ]; then
       if [ "${ENABLE_DISCORD_PUSH}" == "true" ]; then
         log "Pushing image to Discord" "INFO"
-        ${PUSH_PROC_DIR}/push_discord.sh "${IMAGE_FILE_BASE}-122-rectified.jpg" "${push_annotation}" >> $NOAA_LOG 2>&1
+	${PUSH_PROC_DIR}/push_discord.sh "${IMAGE_FILE_BASE}-2-122-rectified.jpg" "${push_annotation}" >> $NOAA_LOG 2>&1
```
This indentation change also looks wrong?

MihajloPi replied:
Also filename changed.

Cadair commented on scripts/receive_meteor.sh:
```diff
@@ -176,59 +176,50 @@ if [ "$METEOR_RECEIVER" == "rtl_fm" ]; then
+    counter=1
 
-    if [ -f "${IMAGE_FILE_BASE}-122-rectified.jpg" ]; then
+    if [ -f "${IMAGE_FILE_BASE}-1-122-rectified.jpg" ]; then
```
Why is this check using a different filename to the one we end up pushing?

MihajloPi replied:
We push every image, I check to see if any image exists by checking if the first image in the series exists. Then, I can choose if I push visible, IR or both images.

Cadair reviewed (commented):
Thanks so much for doing this work, it look great.

I have left a few inline comments, mostly trivial. I do think we need to get the gnuradio stuff working before we could merge this into master.

MihajloPi commented:
Can we create a new development branch so we can merge my commits there for further testing and development?

Cadair commented on install_and_upgrade.sh:
```diff
@@ -71,6 +71,11 @@ else
 fi
 
+log_running "Installing MeteorDemod and OpenCV as its dependency"
+sudo apt install python3-opencv
+sudo apt install ~/raspberry-noaa-v2/software/Meteordemod-2.0.0-Linux.deb
```
These packages should be installed by ansible in this file: https://github.com/jekhokie/raspberry-noaa-v2/blob/master/ansible/roles/common/tasks/dependencies.yml

Cadair commented on ansible/roles/common/tasks/configs.yml:
```diff
@@ -30,7 +30,15 @@
     mode: 0644
-
+    
+- name: MeteorDemod config settings.ini file
```
This failed for me on the first run, with the following error:
```
TASK [common : MeteorDemod config settings.ini file] *********************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "checksum": "b444a9fbcaeebec7550c33a7c619746adef63df7", "msg": "Destination /home/pi/.config/meteordemod not writable"}
        to retry, use: --limit @/home/pi/raspberry-noaa-v2/ansible/site.retry

PLAY RECAP ***********************************************
localhost                  : ok=5    changed=3    unreachable=0    failed=1 
```

I *suspect* what has happened is your deb has added files here (presumably hard coded to the pi user? which is a really bad idea given #461) and therefore they have been created owned by root.

MihajloPi replied:
As I said, I'm not familiar eith Ansible and as a result, you will have to do that on your own. I just created some starting point which I think *could* work.

I used a deb file from MeteorDemod Github, it only works for the user named pi, he still haven't created deb file which coučd work for any user.

Cadair commented:
> Can we create a new development branch so we can merge my commits there for further testing and development?

We could, though we could also just continue to work on it as part of this PR? Do you have specific thoughts on why it would be better PRed off of a branch on the upstream repo?

MihajloPi commented:
It's better to create a development branch which we would constantly work on improving, and IF AND ONLY IF we are satisfied with the result, we would simply merge it with the master branch. That way, we could commit any changes to development branch freely without the fear of destroying the working master branch.

Commit:
Added meteor_demod alias

I've corrected file path for meteor_demod program. It is in `/usr/local/bin`, not in `/usr/bin/` as previously stated.

Commit:
Added underscore back to $METEOR_DEMOD

I made a mistake of removing an underscore. $METEORDEMOD is MeteorDemod by Digitelektro, here we want to use $METEOR_DEMOD which is meteor_demod, a completely different program.

MihajloPi commented:
**IMPORTANT!**

I've made a change in the `common.sh` and `meteor_demodulate_qpsk.sh` files since I didn't receive any Meteor image this evening. I accidentaly removed alias for meteor_demod, and the alias which was previously there was wrong. It pointed to `/usr/bin/meteor_demod` but the program was in `/usr/local/bin/meteor_demod`.
